fuzz/provider.c

```diff
@@ -0,0 +1,687 @@
+
+DEFINE_ALGORITHMS(mac, EVP_MAC)
+
+DEFINE_ALGORITHMS(keymgmt, EVP_KEYMGMT)
+
+DEFINE_ALGORITHMS(signature, EVP_SIGNATURE)
+
+DEFINE_ALGORITHMS(asym_ciphers, EVP_ASYM_CIPHER)
+
+static OSSL_LIB_CTX *libctx = NULL;
+
+int FuzzerInitialize(int *argc, char ***argv)
+{
+    libctx = OSSL_LIB_CTX_new();
+    if (libctx == NULL)
+        return 0;
+
+    init_digests(libctx);
+    init_kdf(libctx);
+    init_cipher(libctx);
+    init_kem(libctx);
+    init_keyexch(libctx);
+    init_rand(libctx);
+    init_mac(libctx);
+    init_keymgmt(libctx);
+    init_signature(libctx);
+    init_asym_ciphers(libctx);
+    return 1;
+}
+
+void FuzzerCleanup(void)
+{
+    cleanup_digests();
+    cleanup_kdf();
+    cleanup_cipher();
+    cleanup_kem();
+    cleanup_keyexch();
+    cleanup_rand();
+    cleanup_mac();
+    cleanup_keymgmt();
+    cleanup_signature();
+    cleanup_asym_ciphers();
+
+    OSSL_LIB_CTX_free(libctx);
+}
+
+static int read_uint(const uint8_t **buf, size_t *len, uint64_t **res)
+{
+    int r = 1;
+
+    if (*len < sizeof(uint64_t)) {
+        r = 0;
+        goto end;
+    }
+
+    *res = malloc(sizeof(uint64_t));
+    **res = (uint64_t) **buf;
+
+    *buf += sizeof(uint64_t);
+    *len -= sizeof(uint64_t);
+end:
+    return r;
+}
+
+static int read_int(const uint8_t **buf, size_t *len, int64_t **res)
+{
+    int r = 1;
+
+    if (*len < sizeof(int64_t)) {
+        r = 0;
+        goto end;
+    }
+
+
+    *res = malloc(sizeof(int64_t));
+    **res = (int64_t) **buf;
+
+    *buf += sizeof(int64_t);
+    *len -= sizeof(int64_t);
+end:
+    return r;
+}
+
+static int read_double(const uint8_t **buf, size_t *len, double **res)
+{
+    int r = 1;
+
+    if (*len < sizeof(double)) {
+        r = 0;
+        goto end;
+    }
+
+    *res = malloc(sizeof(double));
+    **res = (double) **buf;
+
+    *buf += sizeof(double);
+    *len -= sizeof(double);
+end:
+    return r;
+}
+
+static int read_utf8_string(const uint8_t **buf, size_t *len, char **res)
+{
+    int r;
+    size_t i;
+    const uint8_t *ptr = *buf;
+    int found = 0;
+
+    for (i = 0; i < *len; ++i) {
+        if (*ptr == 0) {
+            ptr++;
+            found = 1;
+            break;
+        }
+        ptr++;
+    }
+
+    if (!found) {
+        r = -1;
+        goto end;
+    }
+
+    *res = (char *) *buf;
+
+    r = ptr - *buf;
+    *len -= r;
+    *buf = ptr;
+
+end:
+    return r;
+}
+
+static int read_utf8_ptr(const uint8_t **buf, size_t *len, char **res)
+{
+    if (*len > 0 && **buf == 0xFF) {
+        /* represent NULL somehow */
+        *res = NULL;
+        *buf += 1;
+        *len -= 1;
+        return 0;
+    }
+    return read_utf8_string(buf, len, res);
+}
+
+static int read_octet_string(const uint8_t **buf, size_t *len, char **res)
+{
+    int r;
+    size_t i;
+    const uint8_t *ptr = *buf;
+    int found = 0;
+
+    for (i = 0; i < *len; ++i) {
+        if (*ptr == 0xFF &&
+            (i + 1 < *len && *(ptr + 1) == 0xFF)) {
```

Comment:
Doesn't this always include the 0xFF, 0xFF marker in the resulting octet string? Wouldn't it be better to omit it? Actually I might be tempted to instead use a prefix set of bytes indicating the length to allow 0xFF, 0xFF to be present in the octet string.

Reply:
Essentially yes, separating by magic bytes might cause issues and it is not a clean "serialization" of data.

The idea is to avoid TLV (type-length-value) encoding as getting that right is more difficult for fuzzers. With magic values fuzzers can simply inject and delete data without adjusting the length field.

The downside is that things go wrong if you have e.g. 0xFF in your input data, however that should be negligible.

Are you fine with keeping that for now? There is sadly no real evidence which option is better, however in general the community prefers magic-value separators over TLV encoding.

Reply:
ok - fair enough